internal/streamingstore/streamingstore.go

```diff
@@ -0,0 +1,204 @@
+	carv2 "github.com/ipld/go-car/v2"
+	carstore "github.com/ipld/go-car/v2/storage"
+	"github.com/ipld/go-ipld-prime/storage"
+)
+
+var _ storage.StreamingReadableStorage = (*StreamingStore)(nil)
+var _ storage.ReadableStorage = (*StreamingStore)(nil)
+var _ storage.WritableStorage = (*StreamingStore)(nil)
+
+var errClosed = errors.New("streaming store closed")
+
+// StreamingStore is a storage that opens the underlying CAR storage only when
+// required. So it should avoid creating CAR files that aren't ultimately
+// touched and therefore need cleanup.
+// Two storage systems are used by StreamingStore. One that can perform both
+// read and write operations to serve as the block service for graphsync; and
+// the other that only performs write operations, that can be used to stream
+// CAR contents to the client.
+// Put() operations will write to both storage systems, and will block until
+// both are written.
+type StreamingStore struct {
+	ctx       context.Context
+	roots     []cid.Cid
+	getWriter func() (io.Writer, error)
+	errorCb   func(error)
+
+	lk        sync.Mutex
+	closed    bool
+	f         *os.File
+	readWrite *carstore.StorageCar
+	write     storage.WritableStorage
+}
+
+func NewStreamingStore(ctx context.Context, roots []cid.Cid, getWriter func() (io.Writer, error), errorCb func(error)) *StreamingStore {
+	return &StreamingStore{
+		ctx:       ctx,
+		roots:     roots,
+		getWriter: getWriter,
+		errorCb:   errorCb,
+	}
+}
+
+func (ss *StreamingStore) Has(ctx context.Context, key string) (bool, error) {
+	if ss.ctx.Err() != nil {
+		return false, ss.ctx.Err()
+	}
+	reader, err := ss.lazyReadWrite(true)
+	if err != nil {
+		ss.errorCb(err)
+		return false, err
+	}
+	return reader.Has(ctx, key)
+}
+
+func (ss *StreamingStore) Get(ctx context.Context, key string) ([]byte, error) {
+	if ss.ctx.Err() != nil {
+		return nil, ss.ctx.Err()
+	}
+	reader, err := ss.lazyReadWrite(true)
+	if err != nil {
+		ss.errorCb(err)
+		return nil, err
+	}
+	return reader.Get(ctx, key)
+}
+
+func (ss *StreamingStore) GetStream(ctx context.Context, key string) (io.ReadCloser, error) {
+	if ss.ctx.Err() != nil {
+		return nil, ss.ctx.Err()
+	}
+	reader, err := ss.lazyReadWrite(true)
+	if err != nil {
+		ss.errorCb(err)
+		return nil, err
+	}
+	return reader.GetStream(ctx, key)
+}
+
+func (ss *StreamingStore) Put(ctx context.Context, key string, data []byte) error {
+	if ss.ctx.Err() != nil {
+		return ss.ctx.Err()
+	}
+	writer, err := ss.lazyWriter()
+	if err != nil {
+		ss.errorCb(err)
+		return err
+	}
+	return writer.Put(ctx, key, data)
+}
+
+func (ss *StreamingStore) Close() error {
+	ss.lk.Lock()
+	defer ss.lk.Unlock()
+	// don't need to Finalize the stores because we're writing CARv1
+	ss.closed = true
+	if ss.f != nil {
+		err := ss.f.Close()
+		if err != nil {
+			return err
+		}
+		return os.Remove(ss.f.Name())
+	}
+	return nil
+}
+
+func (ss *StreamingStore) lazyReadWrite(lock bool) (*carstore.StorageCar, error) {
+	if lock {
```

Comment:
generally I'd just make a second lazyReadWriteLock that wraps this method for this.

Reply:
I want to synchronize `closed` too, and have manipulation of that impact on both the readwrite and the write property so I either need one lock or have both and use them both whenever I go near `closed`.

So instead I refactored an unsynchronized `setupReadWrite()` that both can call